File: implementations/ts/src/term-json.ts
// Parse the JSON term profile (ERRATA-2 E1) into Term/Pred. Strings are NFC-normalized at parse
// time so term-side comparisons are NFC-vs-NFC (data strings are NFC by validation, D11).

import type { GroupKey, MaskPolicy, Term } from "./eval.js";
import type { MergeFn, Order, Policy, PropPolicy } from "./policy.js";
import type { Cmp, EntityMatch, PPred, Pred, StrMatch, ValMatch } from "./pred.js";
import type { Primitive } from "./types.js";

const CMPS: readonly Cmp[] = ["eq", "neq", "lt", "lte", "gt", "gte", "prefix", "inSet"];

function nfc(s: string): string {
  return s.normalize("NFC");
}

function asObject(x: unknown, what: string): Record<string, unknown> {
  if (typeof x !== "object" || x === null || Array.isArray(x)) {
    throw new Error(`expected object for ${what}`);
  }
  return x as Record<string, unknown>;
}

function parsePrimitive(v: unknown, what: string): Primitive {
  if (typeof v === "string") return nfc(v);
  if (typeof v === "boolean") return v;
  if (typeof v === "number") {
    if (!Number.isFinite(v)) throw new Error(`${what}: numeric constant must be finite`);
    return v;
  }
  throw new Error(`${what}: constant must be string | number | boolean`);
}

function parseCmp(v: unknown, what: string): Cmp {
  if (typeof v !== "string" || !CMPS.includes(v as Cmp)) {
    throw new Error(`${what}: unknown cmp ${String(v)}`);
  }
  return v as Cmp;
}

function parseStrMatch(raw: unknown, what: string): StrMatch {
  const o = asObject(raw, what);
  if (typeof o["exact"] === "string") return { kind: "exact", value: nfc(o["exact"]) };
  if (typeof o["prefix"] === "string") return { kind: "prefix", value: nfc(o["prefix"]) };
  if (Array.isArray(o["inSet"])) {
    return {
      kind: "inSet",
      values: o["inSet"].map((s) => {
        if (typeof s !== "string") throw new Error(`${what}: inSet members must be strings`);
        return nfc(s);
      }),
    };
  }
  throw new Error(`${what}: StrMatch must be exact | prefix | inSet`);
}

function parseValMatch(raw: unknown, what: string): ValMatch {
  const o = asObject(raw, what);
  if (o["vcmp"] !== undefined) {
    const v = asObject(o["vcmp"], `${what}.vcmp`);
    const cmp = parseCmp(v["cmp"], `${what}.vcmp`);
    if (cmp === "inSet")
      throw new Error(`${what}: vcmp cmp inSet is not allowed; use the inSet arm`);
    const value = parsePrimitive(v["value"], `${what}.vcmp`);
    if (cmp === "prefix" && typeof value !== "string") {
      throw new Error(`${what}: prefix requires a string constant`);
    }
    return { kind: "vcmp", cmp, value };
  }
  if (Array.isArray(o["between"])) {
    if (o["between"].length !== 2) throw new Error(`${what}: between takes [lo, hi]`);
    return {
      kind: "between",
      lo: parsePrimitive(o["between"][0], `${what}.between`),
      hi: parsePrimitive(o["between"][1], `${what}.between`),
    };
  }
  if (Array.isArray(o["inSet"])) {
    return { kind: "inSet", values: o["inSet"].map((v) => parsePrimitive(v, `${what}.inSet`)) };
  }
  throw new Error(`${what}: ValMatch must be vcmp | between | inSet`);
}

function parsePPred(raw: unknown): PPred {
  const o = asObject(raw, "hasPointer");
  const out: {
    role?: StrMatch;
    targetEntity?: EntityMatch;
    targetDelta?: string;
    context?: StrMatch;
    targetIsPrimitive?: boolean;
    targetValue?: ValMatch;
  } = {};
  if (o["role"] !== undefined) out.role = parseStrMatch(o["role"], "hasPointer.role");
  if (o["targetEntity"] !== undefined) {
    const te = o["targetEntity"];
    if (typeof te === "string") {
      out.targetEntity = { kind: "const", id: nfc(te) };
    } else {
      const v = asObject(te, "targetEntity");
      if (v["var"] !== "root") throw new Error('targetEntity must be a string or {var: "root"}');
      out.targetEntity = { kind: "root" };
    }
  }
  if (o["targetDelta"] !== undefined) {
    if (typeof o["targetDelta"] !== "string") throw new Error("targetDelta must be a string");
    out.targetDelta = o["targetDelta"];
  }
  if (o["context"] !== undefined) out.context = parseStrMatch(o["context"], "hasPointer.context");
  if (o["targetIsPrimitive"] !== undefined) {
    if (typeof o["targetIsPrimitive"] !== "boolean") {
      throw new Error("targetIsPrimitive must be a boolean");
    }
    out.targetIsPrimitive = o["targetIsPrimitive"];
  }
  if (o["targetValue"] !== undefined) {
    out.targetValue = parseValMatch(o["targetValue"], "hasPointer.targetValue");
  }
  if (Object.keys(out).length === 0) throw new Error("hasPointer requires at least one field (E1)");
  return out;
}

export function parsePred(raw: unknown): Pred {
  if (raw === "true") return { kind: "true" };
  if (raw === "false") return { kind: "false" };
  const o = asObject(raw, "pred");
  if (o["match"] !== undefined) {
    const m = asObject(o["match"], "match");
    const field = m["field"];
    if (field !== "author" && field !== "timestamp" && field !== "id") {
      throw new Error(`match: unknown field ${String(field)}`);
    }
    const cmp = parseCmp(m["cmp"], "match");
    const rawConst = m["const"];
    const constant =
      cmp === "inSet"
        ? (() => {
            if (!Array.isArray(rawConst)) throw new Error("match: inSet requires an array const");
            return rawConst.map((v) => parsePrimitive(v, "match.const"));
          })()
        : parsePrimitive(rawConst, "match.const");
    if (cmp === "prefix" && typeof constant !== "string") {
      throw new Error("match: prefix requires a string const");
    }
    return { kind: "match", field, cmp, constant };
  }
  if (o["hasPointer"] !== undefined)
    return { kind: "hasPointer", ppred: parsePPred(o["hasPointer"]) };
  if (o["and"] !== undefined || o["or"] !== undefined) {
    const key = o["and"] !== undefined ? "and" : "or";
    const arr = o[key];
    if (!Array.isArray(arr) || arr.length !== 2)
      throw new Error(`${key} takes exactly [Pred, Pred] (E1)`);
    const left = parsePred(arr[0]);
    const right = parsePred(arr[1]);
    return key === "and" ? { kind: "and", left, right } : { kind: "or", left, right };
  }
  if (o["not"] !== undefined) return { kind: "not", pred: parsePred(o["not"]) };
  throw new Error("pred must be true | false | match | hasPointer | and | or | not");
}

function parseMaskPolicy(raw: unknown): MaskPolicy {
  if (raw === "drop") return { kind: "drop" };
  if (raw === "annotate") return { kind: "annotate" };
  const o = asObject(raw, "mask.policy");
  if (o["trust"] !== undefined) return { kind: "trust", pred: parsePred(o["trust"]) };
  throw new Error("mask policy must be drop | annotate | {trust: Pred}");
}

const MERGE_FNS: readonly MergeFn[] = ["max", "min", "sum", "count", "and", "or", "concatSorted"];

function parseOrder(raw: unknown): Order {
  if (raw === "lexById") return { kind: "lexById" };
  const o = asObject(raw, "order");
  if (o["byTimestamp"] !== undefined) {
    if (o["byTimestamp"] !== "desc" && o["byTimestamp"] !== "asc") {
      throw new Error("byTimestamp must be desc | asc");
    }
    return { kind: "byTimestamp", dir: o["byTimestamp"] };
  }
  if (Array.isArray(o["byAuthorRank"])) {
    return {
      kind: "byAuthorRank",
      authors: o["byAuthorRank"].map((a) => {
        if (typeof a !== "string") throw new Error("byAuthorRank entries must be strings");
        return nfc(a);
      }),
    };
  }
  if (o["byPred"] !== undefined) {
    const p = asObject(o["byPred"], "byPred");
    return { kind: "byPred", pred: parsePred(p["pred"]), then: parseOrder(p["then"]) };
  }
  throw new Error("order must be lexById | byTimestamp | byAuthorRank | byPred");
}

function parsePropPolicy(raw: unknown): PropPolicy {
  const o = asObject(raw, "propPolicy");
  if (o["pick"] !== undefined) {
    return { kind: "pick", order: parseOrder(asObject(o["pick"], "pick")["order"]) };
  }
  if (o["all"] !== undefined) {
    return { kind: "all", order: parseOrder(asObject(o["all"], "all")["order"]) };
  }
  if (o["merge"] !== undefined) {
    if (!MERGE_FNS.includes(o["merge"] as MergeFn)) {
      throw new Error("unknown merge fn " + String(o["merge"]));
    }
    return { kind: "merge", fn: o["merge"] as MergeFn };
  }
  if (o["conflicts"] !== undefined) {
    return { kind: "conflicts", order: parseOrder(asObject(o["conflicts"], "conflicts")["order"]) };
  }
  if (o["absentAs"] !== undefined) {
    const a = asObject(o["absentAs"], "absentAs");
    return {
      kind: "absentAs",
      constant: parsePrimitive(a["const"], "absentAs.const"),
      then: parsePropPolicy(a["then"]),
    };
  }
  throw new Error("propPolicy must be pick | all | merge | conflicts | absentAs");
}

export function parsePolicy(raw: unknown): Policy {
  const o = asObject(raw, "policy");
  const props = new Map<string, PropPolicy>();
  if (o["props"] !== undefined) {
    for (const [k, v] of Object.entries(asObject(o["props"], "policy.props"))) {
      props.set(nfc(k), parsePropPolicy(v));
    }
  }
  return { props, default: parsePropPolicy(o["default"]) };
}

function parseGroupKey(raw: unknown): GroupKey {
  if (raw === "byTargetContext") return { kind: "byTargetContext" };
  if (raw === "byRole") return { kind: "byRole" };
  const o = asObject(raw, "group.key");
  if (typeof o["const"] === "string") return { kind: "const", prop: nfc(o["const"]) };
  throw new Error("group key must be byTargetContext | byRole | {const: string}");
}

export function parseTerm(raw: unknown): Term {
  if (raw === "input") return { kind: "input" };
  const o = asObject(raw, "term");
  switch (o["op"]) {
    case "select":
      return { kind: "select", pred: parsePred(o["pred"]), of: parseTerm(o["in"]) };
    case "union":
      return { kind: "union", left: parseTerm(o["left"]), right: parseTerm(o["right"]) };
    case "mask":
      return { kind: "mask", policy: parseMaskPolicy(o["policy"]), of: parseTerm(o["in"]) };
    case "group":
      return { kind: "group", key: parseGroupKey(o["key"]), of: parseTerm(o["in"]) };
    case "expand": {
      if (typeof o["schema"] !== "string") throw new Error("expand.schema must be a string");
      return {
        kind: "expand",
        role: parseStrMatch(o["role"], "expand.role"),
        schema: nfc(o["schema"]),
        of: parseTerm(o["in"]),
      };
    }
    case "fix": {
      if (typeof o["schema"] !== "string") throw new Error("fix.schema must be a string");
      if (typeof o["entity"] !== "string") throw new Error("fix.entity must be a string");
      return { kind: "fix", schema: nfc(o["schema"]), entity: nfc(o["entity"]) };
    }
    case "resolve":
      return { kind: "resolve", policy: parsePolicy(o["policy"]), of: parseTerm(o["in"]) };
    case "prune": {
      const keep = o["keep"] === "all" ? "all" : parseStrMatch(o["keep"], "prune.keep");
      return { kind: "prune", keep, of: parseTerm(o["in"]) };
    }
    default:
      throw new Error(`unknown term op ${String(o["op"])}`);
  }
}

File: implementations/rust/src/term_json.rs
//! Parse the JSON term profile (ERRATA-2 E1) into Term/Pred. Mirrors ../ts/src/term-json.ts.
//! Strings are NFC-normalized at parse time.

use serde_json::Value;
use unicode_normalization::UnicodeNormalization;

use crate::eval::{GroupKey, MaskPolicy, PruneKeep, Term};
use crate::pred::{Cmp, Field, MatchConst, PPred, Pred, StrMatch, ValMatch};
use crate::types::Primitive;

fn nfc(s: &str) -> String {
    s.nfc().collect()
}

fn parse_primitive(v: &Value, what: &str) -> Result<Primitive, String> {
    match v {
        Value::String(s) => Ok(Primitive::Str(nfc(s))),
        Value::Bool(b) => Ok(Primitive::Bool(*b)),
        Value::Number(_) => {
            let n = v.as_f64().ok_or_else(|| format!("{what}: bad number"))?;
            if !n.is_finite() {
                return Err(format!("{what}: numeric constant must be finite"));
            }
            Ok(Primitive::Num(n))
        }
        _ => Err(format!(
            "{what}: constant must be string | number | boolean"
        )),
    }
}

fn parse_cmp(v: &Value, what: &str) -> Result<Cmp, String> {
    match v.as_str() {
        Some("eq") => Ok(Cmp::Eq),
        Some("neq") => Ok(Cmp::Neq),
        Some("lt") => Ok(Cmp::Lt),
        Some("lte") => Ok(Cmp::Lte),
        Some("gt") => Ok(Cmp::Gt),
        Some("gte") => Ok(Cmp::Gte),
        Some("prefix") => Ok(Cmp::Prefix),
        Some("inSet") => Ok(Cmp::InSet),
        _ => Err(format!("{what}: unknown cmp {v}")),
    }
}

fn parse_str_match(v: &Value, what: &str) -> Result<StrMatch, String> {
    let o = v
        .as_object()
        .ok_or_else(|| format!("{what}: expected object"))?;
    if let Some(s) = o.get("exact").and_then(Value::as_str) {
        return Ok(StrMatch::Exact(nfc(s)));
    }
    if let Some(s) = o.get("prefix").and_then(Value::as_str) {
        return Ok(StrMatch::Prefix(nfc(s)));
    }
    if let Some(arr) = o.get("inSet").and_then(Value::as_array) {
        let values = arr
            .iter()
            .map(|s| {
                s.as_str()
                    .map(nfc)
                    .ok_or_else(|| format!("{what}: inSet members must be strings"))
            })
            .collect::<Result<Vec<_>, _>>()?;
        return Ok(StrMatch::InSet(values));
    }
    Err(format!("{what}: StrMatch must be exact | prefix | inSet"))
}

fn parse_val_match(v: &Value, what: &str) -> Result<ValMatch, String> {
    let o = v
        .as_object()
        .ok_or_else(|| format!("{what}: expected object"))?;
    if let Some(vc) = o.get("vcmp") {
        let vo = vc
            .as_object()
            .ok_or_else(|| format!("{what}.vcmp: expected object"))?;
        let cmp = parse_cmp(
            vo.get("cmp").unwrap_or(&Value::Null),
            &format!("{what}.vcmp"),
        )?;
        if cmp == Cmp::InSet {
            return Err(format!(
                "{what}: vcmp cmp inSet is not allowed; use the inSet arm"
            ));
        }
        let value = parse_primitive(
            vo.get("value").unwrap_or(&Value::Null),
            &format!("{what}.vcmp"),
        )?;
        if cmp == Cmp::Prefix && !matches!(value, Primitive::Str(_)) {
            return Err(format!("{what}: prefix requires a string constant"));
        }
        return Ok(ValMatch::Vcmp { cmp, value });
    }
    if let Some(arr) = o.get("between").and_then(Value::as_array) {
        if arr.len() != 2 {
            return Err(format!("{what}: between takes [lo, hi]"));
        }
        return Ok(ValMatch::Between {
            lo: parse_primitive(&arr[0], &format!("{what}.between"))?,
            hi: parse_primitive(&arr[1], &format!("{what}.between"))?,
        });
    }
    if let Some(arr) = o.get("inSet").and_then(Value::as_array) {
        let values = arr
            .iter()
            .map(|x| parse_primitive(x, &format!("{what}.inSet")))
            .collect::<Result<Vec<_>, _>>()?;
        return Ok(ValMatch::InSet(values));
    }
    Err(format!("{what}: ValMatch must be vcmp | between | inSet"))
}

fn parse_ppred(v: &Value) -> Result<PPred, String> {
    let o = v.as_object().ok_or("hasPointer: expected object")?;
    let mut out = PPred::default();
    let mut any = false;
    if let Some(r) = o.get("role") {
        out.role = Some(parse_str_match(r, "hasPointer.role")?);
        any = true;
    }
    if let Some(e) = o.get("targetEntity") {
        out.target_entity = Some(nfc(e.as_str().ok_or("targetEntity must be a string")?));
        any = true;
    }
    if let Some(d) = o.get("targetDelta") {
        out.target_delta = Some(
            d.as_str()
                .ok_or("targetDelta must be a string")?
                .to_string(),
        );
        any = true;
    }
    if let Some(c) = o.get("context") {
        out.context = Some(parse_str_match(c, "hasPointer.context")?);
        any = true;
    }
    if let Some(b) = o.get("targetIsPrimitive") {
        out.target_is_primitive = Some(b.as_bool().ok_or("targetIsPrimitive must be a boolean")?);
        any = true;
    }
    if let Some(tv) = o.get("targetValue") {
        out.target_value = Some(parse_val_match(tv, "hasPointer.targetValue")?);
        any = true;
    }
    if !any {
        return Err("hasPointer requires at least one field (E1)".to_string());
    }
    Ok(out)
}

pub fn parse_pred(raw: &Value) -> Result<Pred, String> {
    if raw == "true" {
        return Ok(Pred::True);
    }
    if raw == "false" {
        return Ok(Pred::False);
    }
    let o = raw.as_object().ok_or("pred: expected object")?;
    if let Some(m) = o.get("match") {
        let mo = m.as_object().ok_or("match: expected object")?;
        let field = match mo.get("field").and_then(Value::as_str) {
            Some("author") => Field::Author,
            Some("timestamp") => Field::Timestamp,
            Some("id") => Field::Id,
            other => return Err(format!("match: unknown field {other:?}")),
        };
        let cmp = parse_cmp(mo.get("cmp").unwrap_or(&Value::Null), "match")?;
        let raw_const = mo.get("const").unwrap_or(&Value::Null);
        let constant = if cmp == Cmp::InSet {
            let arr = raw_const
                .as_array()
                .ok_or("match: inSet requires an array const")?;
            MatchConst::Many(
                arr.iter()
                    .map(|v| parse_primitive(v, "match.const"))
                    .collect::<Result<Vec<_>, _>>()?,
            )
        } else {
            let one = parse_primitive(raw_const, "match.const")?;
            if cmp == Cmp::Prefix && !matches!(one, Primitive::Str(_)) {
                return Err("match: prefix requires a string const".to_string());
            }
            MatchConst::One(one)
        };
        return Ok(Pred::Match {
            field,
            cmp,
            constant,
        });
    }
    if let Some(hp) = o.get("hasPointer") {
        return Ok(Pred::HasPointer(parse_ppred(hp)?));
    }
    for (key, is_and) in [("and", true), ("or", false)] {
        if let Some(arr) = o.get(key) {
            let arr = arr
                .as_array()
                .filter(|a| a.len() == 2)
                .ok_or_else(|| format!("{key} takes exactly [Pred, Pred] (E1)"))?;
            let left = Box::new(parse_pred(&arr[0])?);
            let right = Box::new(parse_pred(&arr[1])?);
            return Ok(if is_and {
                Pred::And(left, right)
            } else {
                Pred::Or(left, right)
            });
        }
    }
    if let Some(n) = o.get("not") {
        return Ok(Pred::Not(Box::new(parse_pred(n)?)));
    }
    Err("pred must be true | false | match | hasPointer | and | or | not".to_string())
}

fn parse_mask_policy(raw: &Value) -> Result<MaskPolicy, String> {
    if raw == "drop" {
        return Ok(MaskPolicy::Drop);
    }
    if raw == "annotate" {
        return Ok(MaskPolicy::Annotate);
    }
    if let Some(o) = raw.as_object() {
        if let Some(p) = o.get("trust") {
            return Ok(MaskPolicy::Trust(parse_pred(p)?));
        }
    }
    Err("mask policy must be drop | annotate | {trust: Pred}".to_string())
}

fn parse_group_key(raw: &Value) -> Result<GroupKey, String> {
    if raw == "byTargetContext" {
        return Ok(GroupKey::ByTargetContext);
    }
    if raw == "byRole" {
        return Ok(GroupKey::ByRole);
    }
    if let Some(o) = raw.as_object() {
        if let Some(s) = o.get("const").and_then(Value::as_str) {
            return Ok(GroupKey::Const(nfc(s)));
        }
    }
    Err("group key must be byTargetContext | byRole | {const: string}".to_string())
}

pub fn parse_term(raw: &Value) -> Result<Term, String> {
    if raw == "input" {
        return Ok(Term::Input);
    }
    let o = raw.as_object().ok_or("term: expected object")?;
    match o.get("op").and_then(Value::as_str) {
        Some("select") => Ok(Term::Select {
            pred: parse_pred(o.get("pred").unwrap_or(&Value::Null))?,
            of: Box::new(parse_term(o.get("in").unwrap_or(&Value::Null))?),
        }),
        Some("union") => Ok(Term::Union {
            left: Box::new(parse_term(o.get("left").unwrap_or(&Value::Null))?),
            right: Box::new(parse_term(o.get("right").unwrap_or(&Value::Null))?),
        }),
        Some("mask") => Ok(Term::Mask {
            policy: parse_mask_policy(o.get("policy").unwrap_or(&Value::Null))?,
            of: Box::new(parse_term(o.get("in").unwrap_or(&Value::Null))?),
        }),
        Some("group") => Ok(Term::Group {
            key: parse_group_key(o.get("key").unwrap_or(&Value::Null))?,
            of: Box::new(parse_term(o.get("in").unwrap_or(&Value::Null))?),
        }),
        Some("prune") => {
            let keep_raw = o.get("keep").unwrap_or(&Value::Null);
            let keep = if keep_raw == "all" {
                PruneKeep::All
            } else {
                PruneKeep::Match(parse_str_match(keep_raw, "prune.keep")?)
            };
            Ok(Term::Prune {
                keep,
                of: Box::new(parse_term(o.get("in").unwrap_or(&Value::Null))?),
            })
        }
        other => Err(format!("unknown term op {other:?}")),
    }
}
